### PR TITLE
Small rope improvements

### PR DESCRIPTION
**Pull request by jridgewell**
The main improvement is the addition of an `uncommitted_static` reference, which attempts to prioritize using a static `Bytes` over an extremely short owned `Bytes`. When a `Bytes` references static memory, clones of the `Bytes` is extremely cheap, but when it's non-static it needs to reference count.

### CODE REVIEW

File: crates/turbo-tasks-fs/src/rope.rs [hunk 1]
```diff
@@ -157,26 +181,31 @@ impl RopeBuilder {
     ///
     /// This may be called multiple times without issue.
     pub fn finish(&mut self) {
-        if !self.writable.is_empty() {
+        if let Some(uncommited_static) = self.uncommited_static.take() {
```
Comment (sokra) [inline crates/turbo-tasks-fs/src/rope.rs:184]: This works with the invariant of uncommitted_static and writable are exclusive.

Maybe encode this into an enum:

``` rust
enum Uncommitted {
  None,
  Static(&'static [u8])
  Owned(Vec<u8>)
}
```

Comment (sokra) [reply crates/turbo-tasks-fs/src/rope.rs:184]: Or without None and `Option<Uncommitted>`

**Review by sokra — CHANGES_REQUESTED** [review]

File: crates/turbo-tasks-fs/src/rope.rs [hunk 2]
```diff
@@ -131,6 +143,18 @@ impl RopeBuilder {
...
         // that buffer when more bytes are pushed.
         if bytes.len() < mem::size_of::<Bytes>() {
+            // If we've not already pushed static bytes, we attempt to store the bytes for
+            // later. If we push owned bytes or another static bytes, then this attempt will
+            // fail and we'll instead concatenate into a single owned Bytes. But if we don't
+            // push anything (build the Rope), or concatenate another Rope (we can't join
+            // our bytes with the InnerRope of another Rope), we'll be able to commit a
+            // static Bytes reference and save overall memory (a small static Bytes
+            // reference is better than a small owned Bytes reference).
+            if self.uncommited_static.is_none() {
```
[the remark this answers is not in the document]
Comment (jridgewell) [reply crates/turbo-tasks-fs/src/rope.rs]: Done.

File: crates/turbo-tasks-fs/src/rope.rs (hunk 1 above)
Comment (jridgewell) [reply crates/turbo-tasks-fs/src/rope.rs]: Done.

**Review by sokra — APPROVED** [review]

**Pull request merged**